include/gsl/gsl_assert

```diff
@@ -72,16 +72,49 @@ struct fail_fast : public std::runtime_error
 {
     explicit fail_fast(char const* const message) : std::runtime_error(message) {}
 };
-}
+namespace details
+{
+template <typename Exception>
+[[noreturn]] void throw_exception(Exception&& exception);
```

Comment:
I'm happy with `throw_exception()` throwing an exception under `GSL_UNENFORCED_ON_CONTRACT_VIOLATION`. That way we can use it for cases like `narrow()` where the only options a user gets are: exception or terminate (and we are allowing people to choose terminate because it is strictly more severe and allows non-exception environments). 

Saying "no contract enforcement" should apply only to contract enforcement, not to a function like `narrow()`, whose purpose for existence is to always check-and-ensure.